numpy/lib/function_base.py

```diff
@@ -1595,22 +1622,62 @@ def trim_zeros(filt, trim='fb'):
     [1, 2]
 
     """
-    first = 0
-    trim = trim.upper()
-    if 'F' in trim:
-        for i in filt:
-            if i != 0.:
-                break
-            else:
-                first = first + 1
-    last = len(filt)
-    if 'B' in trim:
-        for i in filt[::-1]:
-            if i != 0.:
-                break
-            else:
-                last = last - 1
-    return filt[first:last]
+    trim = trim.lower()
+
+    if axis is None:
+        # Apply iteratively to all axes
+        axis = range(filt.ndim)
+
+    # Normalize axes to 1D-array
+    axis = np.asarray(axis, dtype=np.intp)
+    if axis.ndim == 0:
+        axis = np.asarray([axis], dtype=np.intp)
+
+    absolutes = np.abs(filt)
+    lengths = np.zeros((absolutes.ndim, 2), dtype=np.intp)
+
+    for current_axis in axis:
+        absolutes.take([], current_axis)  # Raises if axis is out of bounds
+        if current_axis < 0:
+            current_axis += absolutes.ndim
+
+        # Reduce to envelope along all axes except the selected one
+        reduced = np.moveaxis(absolutes, current_axis, -1)
+        for _ in range(absolutes.ndim - 1):
+            reduced = reduced.max(axis=0)
+        assert reduced.ndim == 1
+
+        if atol > 0:
+            reduced[reduced <= atol] = 0
+        if rtol > 0:
+            reduced[reduced <= rtol * reduced.max()] = 0
+
+        # Find start and stop indices for current dimension
+        start, stop = np.nonzero(reduced)[0][[0, -1]]
+        stop += 1
```

Comment:
These two lines replace the old behavior that would search for `start` and `stop` with two simple for loops. I haven't benchmarked it yet but will do so. 

However, as this is likely the bottleneck of the function, I'd like to do this (as well as the reduction to the absolute envelope above) in Cython (or C). Would that be appreciated or is the lack of compiled code in the lib module deliberate?